File: BTLib/src/com/offsetnull/bt/trigger/TriggerData.java
package com.offsetnull.bt.trigger;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.offsetnull.bt.responder.TriggerResponder;
import com.offsetnull.bt.responder.ack.AckResponder;
import com.offsetnull.bt.responder.color.ColorAction;
import com.offsetnull.bt.responder.gag.GagAction;
import com.offsetnull.bt.responder.notification.NotificationResponder;
import com.offsetnull.bt.responder.replace.ReplaceResponder;
import com.offsetnull.bt.responder.script.ScriptResponder;
import com.offsetnull.bt.responder.toast.ToastResponder;

public class TriggerData {

	private String name;
	private String pattern;
	private boolean interpretAsRegex;
	private boolean fireOnce;

	private boolean fired = false;

	private boolean hidden = false;

	private boolean enabled = true;

	private boolean save = true;
	private int sequence = DEFAULT_SEQUENCE;
	private boolean keepEvaluating = DEFAULT_KEEPEVAL;
	private String group = DEFAULT_GROUP;
	private List<TriggerResponder> responders;

	private Pattern p = null;
	private Matcher m = null;
	//private Matcher m = null;

	public TriggerData() {
		name = "";
		pattern = "";
		interpretAsRegex = false;
		responders = new ArrayList<TriggerResponder>();
		fireOnce = false;
		hidden = false;
		enabled = true;
		sequence = DEFAULT_SEQUENCE;
		group = DEFAULT_GROUP;
		keepEvaluating = DEFAULT_KEEPEVAL;
		buildData();
	}

	public TriggerData copy() {
		TriggerData tmp = new TriggerData();
		tmp.name = this.name;
		tmp.pattern = this.pattern;
		tmp.interpretAsRegex = this.interpretAsRegex;
		tmp.fireOnce = this.fireOnce;
		tmp.hidden = this.hidden;
		tmp.enabled = this.enabled;
		tmp.sequence = this.sequence;
		tmp.group = this.group;
		tmp.keepEvaluating = this.keepEvaluating;
		for(TriggerResponder responder : this.responders) {
			tmp.responders.add(responder.copy());
		}
		tmp.buildData();
		return tmp;
	}

	private void buildData() {
		//if(p == null || p.equals("")) return;
		if(this.interpretAsRegex) {
			this.p = Pattern.compile(pattern);
		} else {
			this.p = Pattern.compile("\\Q"+pattern+"\\E");
		}
		this.m = p.matcher("");
	}

	public Matcher getMatcher() {
		return m;
	}

	public boolean equals(Object o) {
		if(o == this) return true;
		if(!(o instanceof TriggerData)) return false;
		TriggerData test = (TriggerData)o;
		if(!test.name.equals(this.name)) return false;
		if(!test.pattern.equals(this.pattern)) return false;
		if(test.interpretAsRegex != this.interpretAsRegex) return false;
		if(test.fireOnce != this.fireOnce) return false;
		if(test.hidden != this.hidden) return false;
		if(test.enabled != this.enabled) return false;
		if(test.sequence != this.sequence) return false;
		if(test.group != this.group) return false;
		if(test.keepEvaluating != this.keepEvaluating) return false;
		Iterator<TriggerResponder> test_responders = test.responders.iterator();
		Iterator<TriggerResponder> my_responders = this.responders.iterator();
		while(test_responders.hasNext()) {
			TriggerResponder test_responder = test_responders.next();
			TriggerResponder my_responder = my_responders.next();

			if(!test_responder.equals(my_responder)) return false;
		}

		return true;
	}

	public static final int DEFAULT_SEQUENCE = 10;
	public static final String DEFAULT_GROUP = "";
	public static final boolean DEFAULT_KEEPEVAL = true;

	public void setName(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	public void setPattern(String pattern) {
		this.pattern = pattern;
		buildData();
	}

	public String getPattern() {
		return pattern;
	}

	public void setInterpretAsRegex(boolean interpretAsRegex) {
		this.interpretAsRegex = interpretAsRegex;
		//buildData();
	}

	public boolean isInterpretAsRegex() {
		return interpretAsRegex;
	}

	public void setResponders(List<TriggerResponder> responders) {
		this.responders = responders;
	}

	public List<TriggerResponder> getResponders() {
		return responders;
	}

	public void setFireOnce(boolean fireOnce) {
		this.fireOnce = fireOnce;
	}

	public boolean isFireOnce() {
		return fireOnce;
	}

	public void setFired(boolean fired) {
		this.fired = fired;
	}

	public boolean isFired() {
		return fired;
	}

	public void setHidden(boolean hidden) {
		this.hidden = hidden;
	}

	public boolean isHidden() {
		return hidden;
	}

	public void setEnabled(boolean enabled) {
		this.enabled = enabled;
	}

	public boolean isEnabled() {
		return enabled;
	}

	public Pattern getCompiledPattern() {
		return p;
	}
	//public boolean matches(CharSequence text) {
		//m.reset(text);

	public void setSave(boolean save) {
		this.save = save;
	}

	public boolean isSave() {
		return save;
	}

	public int getSequence() {
		return sequence;
	}

	public void setSequence(int sequence) {
		this.sequence = sequence;
	}

	public boolean isKeepEvaluating() {
		return keepEvaluating;
	}

	public void setKeepEvaluating(boolean keepEvaluating) {
		this.keepEvaluating = keepEvaluating;
	}

	public String getGroup() {
		return group;
	}

	public void setGroup(String group) {
		this.group = group;
	}

	//}

}

File: lib/src/main/kotlin/org/ncmud/mudwammer/trigger/TriggerData.kt
package org.ncmud.mudwammer.trigger

import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import org.ncmud.mudwammer.responder.TriggerResponder
import java.util.regex.Matcher
import java.util.regex.Pattern

@Entity(tableName = "triggers")
class TriggerData(
    @PrimaryKey(autoGenerate = true) var id: Long = 0,
    var connectionId: Long = 0,
    var name: String = "",
    pattern: String = "",
    @get:JvmName("isInterpretAsRegex") var interpretAsRegex: Boolean = false,
    @get:JvmName("isFireOnce") var fireOnce: Boolean = false,
    @get:JvmName("isHidden") var hidden: Boolean = false,
    @get:JvmName("isEnabled") var enabled: Boolean = true,
    @get:JvmName("isSave") var save: Boolean = true,
    var sequence: Int = DEFAULT_SEQUENCE,
    @get:JvmName("isKeepEvaluating") var keepEvaluating: Boolean = DEFAULT_KEEPEVAL,
    var group: String = DEFAULT_GROUP,
) {
    var pattern: String = pattern
        set(value) { field = value; buildData() }

    @Ignore
    @get:JvmName("isFired")
    var fired: Boolean = false

    @Ignore var responders: MutableList<TriggerResponder> = mutableListOf()

    @Ignore private var p: Pattern? = null
    @Ignore private var m: Matcher? = null

    init { buildData() }

    private fun buildData() {
        p = if (interpretAsRegex) Pattern.compile(pattern) else Pattern.compile("\\Q$pattern\\E")
        m = p!!.matcher("")
    }

    val matcher: Matcher? get() = m
    val compiledPattern: Pattern? get() = p

    fun copy(): TriggerData {
        val tmp = TriggerData(
            name = name, pattern = pattern, interpretAsRegex = interpretAsRegex,
            fireOnce = fireOnce, hidden = hidden, enabled = enabled, save = save,
            sequence = sequence, keepEvaluating = keepEvaluating, group = group
        )
        for (r in responders) { tmp.responders.add(r.copy()) }
        return tmp
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TriggerData) return false
        return name == other.name && pattern == other.pattern
            && interpretAsRegex == other.interpretAsRegex && fireOnce == other.fireOnce
            && hidden == other.hidden && enabled == other.enabled
            && sequence == other.sequence && group == other.group
            && keepEvaluating == other.keepEvaluating
            && responders == other.responders
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + pattern.hashCode()
        result = 31 * result + sequence
        return result
    }

    companion object {
        const val DEFAULT_SEQUENCE = 10
        const val DEFAULT_GROUP = ""
        const val DEFAULT_KEEPEVAL = true
    }
}
